### DISCUSSION TITLE
Build menu hierarchy at run time with a different param list for each command

### DISCUSSION THREAD

**Discussion by tappara** [daniele77/cli#207] [Q&A]
Ciao Daniele,
Nice work, congrats. I like both the design and the implementation. Very neat and efficient.

There is just one thing I'd like you to confirm (or confute).
If I am not mistaken, the function that parses the command params is defined at compile time (as a template parameter pack). 

I have in mind a possible usage of your library where the menu hierarchy is created at run time based on a some sort of other input, for example a JSON file that describes both the commands and the menu hierarchy. Now the problem that I foresee is that each command has an its own different "signature" (ie. list of parameters) which is only known once the input file is parsed and that does not work with the design you chose (template parameter pack).

I can certainly use the version of the parser callback that takes a vector of strings and do the checks there for each different command, but in that case we lose the capability of autocompletion only for a certain number of params (e.g. if you register as parser a function that takes two params - say a int and a string - and the use inputs both of them and then press tab, the cursor does not move, which is a nice way to let the user know that the required input is over).

Am I mistaken or, because of the design choice that you make, no workaround is possible to register a callback whose signature is known only at run time ?
Do you have any other trick to address my scenario?

Thanks in advance for your answer.
N.

**Comment by daniele77** [comment]
Ciao Nicola,

Firstly, I sincerely appreciate your insightful question. Your input could prove beneficial not only for your specific use case but also for other users who encounter similar scenarios.

I will now address your points individually:

> If I am not mistaken, the function that parses the command params is defined at compile time (as a template parameter pack).

Indeed, your understanding is accurate. The template `class VariadicFunctionCommand` is instantiated for each command insertion, with the `struct Select` enabling the creation of a compile-time parser specific to the argument types.

> I have in mind a possible usage of your library where the menu hierarchy is created at run time based on a some sort of other input, for example a JSON file that describes both the commands and the menu hierarchy. Now the problem that I foresee is that each command has an its own different "signature" (ie. list of parameters) which is only known once the input file is parsed and that does not work with the design you chose (template parameter pack).

Indeed, if you intend to construct the command hierarchy dynamically at runtime, the "typed" Insert method (which constructs an internal `VariadicFunctionCommand` object) will not be suitable due to its compile-time nature, orchestrated through templates.

> I can certainly use the version of the parser callback that takes a vector of strings and do the checks there for each different command, but in that case we lose the capability of autocompletion only for a certain number of params (e.g. if you register as parser a function that takes two params - say a int and a string - and the use inputs both of them and then press tab, the cursor does not move, which is a nice way to let the user know that the required input is over).

It's crucial to clarify that the library's autocompletion feature isn't as sophisticated as your expectations suggest. Whether you utilize the `Insert` method with explicit parameter callbacks or opt for the variant accepting `std::vector<std::string>`, autocompletion solely applies to command names, not parameters. Thus, pressing the tab key while inputting parameters won't yield any response, regardless of the quantity entered. Please note that there are plans for enhancing argument completions (as evident in #68 and #157).
Consequently, utilizing the Insert variant that uses a callback accepting `std::vector<std::string>` for your task wouldn't result in the loss of any features.

> Am I mistaken or, because of the design choice that you make, no workaround is possible to register a callback whose signature is known only at run time ? Do you have any other trick to address my scenario?

Given that the current library version does not facilitate autocompletion for command arguments, I propose creating your own `JsonMenuDescriptor`. This descriptor would parse JSON, and for every specified command, invoke `Menu::Insert` with a callback accommodating `std::vector<std::string>` parameters, thereby conducting argument parsing in accordance with the JSON specifications.

As a matter of curiosity, how do you envision specifying the callback command within the JSON structure?

I would greatly appreciate your thoughts on this proposed solution. I am committed to collaborating with you to arrive at the most effective resolution, given the universal applicability of this concern. In the event we formulate a "generic solution," it could serve as a valuable addition to the library's examples.

Looking forward to your input.

Best regards,
Daniele

**Reply by tappara** [reply]
Hi Daniele, 

thanks for your answer and your availability to discuss.
I was clearly mistaken on the autocompletion part, I apologize.

For what concern your proposal of extending your library with a object to parse JSON, I need to spend more time and better understand your design before making any proposal.
Let me churn this into my head a little, hopefully I will come back with some ideas (soon?).

Ciao, 
N.
